ScrollSpy: improve active feedback

guilhas07 opened the pull request:
### Motivation & Context

<!-- Why is this change required? What problem does it solve? -->
Current ScrollSpy behavior had some shortcomings when it comes to detecting the correct active class. 
- Sometimes none of the classes are active
- After clicking to go to a section, it immediately shows as non active

### Solution

~~To fix this I simplified the current logic, by adding an active state for all sections and keeping track of the previous highlighted section. The current rational is to highlight the first section that is intersecting at all times.~~

Different logic for scrolling up or down ~~and custom thresholds are~~ **is** removed due to not being needed anymore.

**Edit**: My previous implementation only considered header elements, which are small due to not wrapping the text in their section. However, when using entire div elements, it became impossible to activate the next element because, in many cases, two elements would be visible simultaneously. To address this, I now activate the element with the larger intersection ratio. To achieve this, I keep track of each element's intersection ratio and have increased the thresholds to recalculate the intersection ratios at every 10% increment.

### Changes
#### Scroll Behaviour:
- Before ( Note that sometimes the active class disappears and other times it jumps to early ):

https://github.com/user-attachments/assets/6db60726-699f-43fa-847b-52e28d59acc2

- After:

https://github.com/user-attachments/assets/3c060388-9749-457d-971d-05b6e8b7cedd

#### Click Behaviour:
- Before (Note the section clicked isn't active after auto scroll)

https://github.com/user-attachments/assets/ca1f2586-3c90-4e03-8955-c2ed4badcdf3

- After

https://github.com/user-attachments/assets/78e22712-3a31-4188-b53d-08851885e5c4

### Type of changes

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Refactoring (non-breaking change)

### Checklist

<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the [contributing guidelines](https://github.com/twbs/bootstrap/blob/main/.github/CONTRIBUTING.md)
- [x] My code follows the code style of the project _(using `npm run lint`)_
- [ ] My change introduces changes to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

### Live Previews

https://deploy-preview-41016--twbs-bootstrap.netlify.app/docs/5.3/getting-started/introduction/

julien-deramond commented:
Sorry for the delay, @guilhas07, and thank you for taking the time to work on this PR!
I haven't digged into the history of the implementation of the Scrollspy component yet, nor looked at your code.
Here's my feedback after testing and comparing the behavior with the current implementation of the Scrollspy component.


- **General improvement:** Overall, this feels less buggy, as all items are active while scrolling. This is a great enhancement!
- **Behavior on specific pages:** On the [Reboot documentation page](https://getbootstrap.com/docs/5.3/content/reboot/), for example, the last element is active in the current implementation but not in your branch. Personally, I don’t mind this change—it aligns with how some other frameworks (e.g., Astro/Starlight) handle small sections that "stick" to the bottom of the page.
- In the [Scrollspy Navbar example from your branch](https://deploy-preview-41016--twbs-bootstrap.netlify.app/docs/5.3/components/scrollspy/#navbar), selecting an item from the dropdown and then scrolling results in two `.active` elements simultaneously (which shouldn't happen IMO).

![Screenshot 2024-12-18 at 08 53 13](https://github.com/user-attachments/assets/ce2c20a1-1dbc-4f14-9113-688b205767f7)

- Possibly linked to one of my previous comments, in the [Scrollspy Nested Nav example from your branch](https://deploy-preview-41016--twbs-bootstrap.netlify.app/docs/5.3/components/scrollspy/#nested-nav), the last element (3.2) never becomes active, even when clicked.

![Screenshot 2024-12-18 at 08 55 30](https://github.com/user-attachments/assets/66beb6da-8739-40d0-8b8f-641509a1daf9)

To wrap up, I think this PR has potential, but we need to address these edge cases. Once finalized, it would be great to add strict unit tests to enforce the new behavior. These tests could also act as documentation for the expected behavior, helping prevent regressions in the future.

Anyone interested in this topic, feel free to jump in and share your feedback! (/cc @twbs/js-review, you have more expertise on this than I do).

guilhas07 commented:
First of all thank you for your feedback!

>General improvement: Overall, this feels less buggy, as all items are active while scrolling. This is a great enhancement!

Thank you!

>Behavior on specific pages: On the [Reboot documentation page](https://getbootstrap.com/docs/5.3/content/reboot/), for example, the last element is active in the current implementation but not in your branch. Personally, I don’t mind this change—it aligns with how some other frameworks (e.g., Astro/Starlight) handle small sections that "stick" to the bottom of the page.

I'm afraid as you already mentioned that is hard to do because technically there are two elements fully shown.

> In the [Scrollspy Navbar example from your branch](https://deploy-preview-41016--twbs-bootstrap.netlify.app/docs/5.3/components/scrollspy/#navbar), selecting an item from the dropdown and then scrolling results in two .active elements simultaneously (which shouldn't happen IMO).

Fixed.

> Possibly linked to one of my previous comments, in the [Scrollspy Nested Nav example from your branch](https://deploy-preview-41016--twbs-bootstrap.netlify.app/docs/5.3/components/scrollspy/#nested-nav), the last element (3.2) never becomes active, even when clicked.

Also fixed but this led me to change the current implementation.

>To wrap up, I think this PR has potential, but we need to address these edge cases. Once finalized, it would be great to add strict unit tests to enforce the new behavior. These tests could also act as documentation for the expected behavior, helping prevent regressions in the future.
Edit: be careful, the tests fail right now. You can run npm run js-test locally to replicate the issues.

I'm looking into it. I've modified some tests to align with the new behavior, and I will try to come up with new ones (if you have any suggestions, I'm all ears). Currently, there is one test failing: `ScrollSpy constructor should not process data if the activeTarget is the same as the given target`. Do we need to keep this behavior?

### Changes:

My previous implementation only considered `header` elements, which are small due to not wrapping the text in their section. However, when using entire div elements, it became impossible to activate the next element because, in many cases, two elements would be visible simultaneously. To address this, I now activate the element with the larger intersection ratio. To achieve this, I keep track of each element's intersection ratio and have increased the thresholds to recalculate the intersection ratios at every 10% increment.

Manchmal1 reviewed (commented):
Danke für die Mühe und ganze Arbeit 👍👍

codyrobbins commented:
Thanks for taking a stab at finally fixing this! In case it’s helpful to cross-reference, there’s a long thread at #36431 discussing various aspects of the current buggy behavior.

Manchmal1 commented:
Ost es aktiv was muss ich machen um es zu aktivieren das es auf meinem zandy oder tv läuft

guilhas07 committed:
fix: scroll to the correct element when smooth scroll is disabled

guilhas07 committed:
Scrollspy: improve active behaviour

guilhas07 committed:
fix: set element to active after a smooth scroll

guilhas07 commented:
@julien-deramond Hey! I've pushed some changes and separated them into commits for easier review. When you have a moment, PTAL.

@codyrobbins Thanks for linking the issue! Would you mind testing this version and sharing your thoughts on it?

P.s.:
I found the wrong `smoothscroll` behaviour in my implementation going over the local website build, but couldn't recreate a test due to being dependent on multiple factors. I think the gist is that `smoothscroll` behaviour, unlike `auto`, doesn't scroll the correct amount of pixels which ends up not triggering the `intersection observer API`.

DenisLopatin commented:
Hi. Will this close #36431? Or is it an offshoot of it, which doesn't fix everything?

mdo commented:
Put some time in with Claude to see if we can get a rewrite in for v6—this wouldn't be able to land in v5 unfortunately. See #42557 for that. Happy to review any feedback there, or adjust as needed. Let me know what you think!

In the mean time, I'll leave this open, but I don't expect to see it merged at this point for v5.

DenisLopatin commented:
> Put some time in with Claude to see if we can get a rewrite in for v6—this wouldn't be able to land in v5 unfortunately. See #42557 for that. Happy to review any feedback there, or adjust as needed. Let me know what you think!
> 
> In the mean time, I'll leave this open, but I don't expect to see it merged at this point for v5.

I have already rewrote the component, keeping the basic functionality. I rewrote all the tests and spent a week on it (some tests were just for show), but they didn't review this commit because there were too many changes. However, it is necessary to rewrite the module. Maybe we should give this commit a chance? Or at least lean on it, or do you see it somehow differently? #41914
